mindmeld/text_preparation/text_preparation_pipeline.py

```diff
@@ -0,0 +1,530 @@
+        Args:
+            text (str): Input text.
+        Returns:
+            normalized_text (str): Normalized Text.
+            forward_map (Dict): Mapping from raw text to modified text.
+            backward_map (Dict): Reverse mapping from modified text to raw text.
+        """
+        if not self.preprocessors:
+            logger.warning(
+                "This instance of TextPreparationPipeline does not have preprocessors."
+                "Returning the original text."
+            )
+            return text
+
+        preprocessed_text = text
+        for preprocessor in self.preprocessors:
+            preprocessed_text = (
+                TextPreparationPipeline.modify_around_mindmeld_annotations(
+                    text=preprocessed_text, function=preprocessor.process
+                )
+            )
+        return preprocessed_text
+
+    def normalize(self, text):
+        """Normalize Text.
+        Args:
+            text (str): Text to normalize.
+        Returns:
+            normalized_text (str): Normalized text.
+        """
+        normalized_tokens = self.get_normalized_tokens_from_text(text)
+        normalized_text = " ".join([t["entity"] for t in normalized_tokens])
+        return normalized_text
+
+    def get_normalized_tokens_from_text(self, text):
+        """
+        Args:
+            text (str): Text to normalize.
+        Returns:
+            normalized_tokens (List[Dict]): Normalized tokens represented as dictionaries.
+                For Example:
+                    norm_token = {
+                        "entity": "order",
+                        "raw_entity": "order",
+                        "raw_token_index": 1,
+                        "raw_start": 1
+                    }
+        """
+        raw_tokens = self.tokenizer.tokenize(text)
+        return self._normalize_tokens(raw_tokens)
+
+    def _normalize_tokens(self, raw_tokens):
+        """Normalize individual token dicts produced by Tokenizers.
+        Args:
+            raw_tokens (List(Dict)): List of raw tokens represented as dictionaries.
+        Returns:
+            normalized_tokens (List[Dict]): Normalized tokens represented as dictionaries.
+                For Example:
+                    norm_token = {
+                        "entity": "order",
+                        "raw_entity": "order",
+                        "raw_token_index": 1,
+                        "raw_start": 1
+                    }`
+        """
+        if not self.normalizers:
+            logger.warning(
+                "This instance of TextPreparationPipeline does not have normalizers."
+                "Returning raw tokens."
+            )
+            return raw_tokens
+
+        normalized_tokens = []
+        for i, raw_token in enumerate(raw_tokens):
+            if not raw_token["text"] or len(raw_token["text"]) == 0:
+                continue
+            normalized_text = self._normalize_single_token(raw_token["text"])
```

Comment:
Maybe consider getting ``normalized_tokens`` as an output of this function directly instead of text? I.e. ``return normalized_text.split()`` from the function?

Reply:
I see what you mean, yes, this would be cleaner! To take it a step further, we can better generalize the post-norm tokenization by using the language specific tokenization (self.tokenize) here

An example of post-norm tokenization is “o’clock” -> “o clock” -> [“o”, “clock”]. In English, the .split() works fine since words are separated by space, however, to generalize better for other languages (like Japanese), so it'd make sense to do self.tokenize(norm_text_single_token) instead of .split().